rfd/0186-optionally-require-reason-for-access-request.md

```diff
@@ -0,0 +1,359 @@
+> role.options.request_access is a global configuration, therefore it doesn't
+> meet the requirement of requiring reason only for _some_ resources/roles.
+
+It has to be noted that role.options.request_access has a global scope. I.e. if
+it's set on any role all access requests to all roles and resources are
+affected.
+
+Example:
+
+For two roles `kube-access-requester` and `node-access-requester` as defined
+below:
+
+```yaml
+kind: role
+version: v7
+metadata:
+  name: kube-access-requester
+spec:
+  allow:
+    request:
+      roles:
+        - kube-access
+  options:
+    request_access: reason-required
+```
+
+```yaml
+kind: role
+version: v7
+metadata:
+  name: node-access-requester
+spec:
+  allow:
+    request:
+      roles:
+        - node-access
+```
+
+A user with both roles `kube-access-requester` and `node-access-requester`
+assigned will be affected by request_access set in role `kube-access-requester`
+while requesting access to role `node-access` allowed by
+`node-access-requester` role.
+
+Because request_access option is global a priority has to be establish when
+more than one role has the option specified:
+
+- reason
+- always
+- reason-required
+- optional (default)
+
+Problems:
+
+- It may be confusing/undesired that roles with request_access option set to
+  "reason-required" affects other roles.
+
+
+#### 2. Add a new value to role.options.request_access and change the option to be scoped
+
+> [!NOTE]
+> This is a breaking change.
+
+The solution would be like in 1. but with changed semantics of request_access
+option. It would become role-scoped instead of global.
+
+Example:
+
+For two roles `kube-access-requester` and `node-access-requester` as defined
+below:
+
+```yaml
+kind: role
+version: v7
+metadata:
+  name: kube-access-requester
+spec:
+  allow:
+    request:
+      roles:
+        - kube-access
+  options:
+    request_access: reason-required
+```
+
+```yaml
+kind: role
+version: v7
+metadata:
+  name: node-access-requester
+spec:
+  allow:
+    request:
+      roles:
+        - node-access
+  options:
+    request_access: optional # default, doesn't need to be set explicitly
+```
+
+A user with `kube-access-requester` and `node-access-requester` roles assigned
+would have to provide a reason while creating an access request for
+`kube-access` role but wouldn't have to provide reason while requesting access
+to `node-access` role.
+
+The example above is clear and simple, but there are problems with the
+implementation:
+
+- **This is a breaking change.** It may come as a surprise to the existing
+  users that only _some_ instead of _all_ roles are automatically requested on
+  login when request_access is set to "always" or "reason".
+- What happens when there are 2 or more roles allowing requesting the same
+  role/resource but they have different request_access set? Should the solution
+  fall back to priorities, like in 1.?
+
+
+#### 3. Add a new role.spec.allow.request.reason.required setting
```

Comment:
Since this is the approach we're taking, let's move it to the top after the UX section, and move the other ones under "Alternatives evaluated" section or something like that. Otherwise it is not clear which one we're implementing.